### ISSUE TITLE
IPython doc strings, question mark does not work

### ISSUE THREAD

**Issue by honeyext** [Microsoft/PTVS#577]
It seems, that IPython interactive window does not support printing doc strings via "?", for example in normal ipython following statement:
```
>>numpy.abs?
```
is equivalent to:
```
>>print numpy.abs.__doc__
```
and double question mark "??" shows the source code if it is available. Are there any reasons for absence of this feature?

PTVS 2.2 RC2 
VS 2013 Express for Desktop
IPython 3.2.0

**Comment by zooba** [comment]
You probably need to configure your interactive window to use IPython mode (either use "Configure interactive window" in the Python Environments window, or through Tools, Options, Python Tools).

Unless you do this, you just get regular Python, which does not support the question mark syntax.

**Comment by honeyext** [comment]
I am in IPython mode, I can use magic functions like %timeit or %cd.

**Comment by zooba** [comment]
Oh, very interesting. I don't have any idea what's happened here. It looks like something changed in IPython 3.0 that broke this on us - IPython 2.4.1 still works fine.

**Comment by honeyext** [comment]
Thank you for your quick response.

**Comment by zooba** [comment]
It looks like IPython 3 started using a full response object for help text, but our wrapper is only expecting a plain response. They're also sending it through the notification channel rather than the IO channel, but I think they also removed that distinction and we still have it so we can easily support older versions.

The fix/workaround is to find `visualstudio_ipython_repl.py` in your [install directory]() and replace the `handle_execute_reply` function (first in the `VsShellChannel` class) with this code:

```python
class VsShellChannel(DefaultHandler, ShellChannel):
    
    def handle_execute_reply(self, content):
        # we could have a payload here...
        payload = content['payload']
        
        for item in payload:
            data = item.get('data')
            if data is not None:
                try:
                    self._vs_backend.km.iopub_channel.write_data
                except AttributeError:
                    pass
                else:
                    self._vs_backend.km.iopub_channel.write_data(data)
                    continue

            output = item.get('text', None)
            if output is not None:
                self._vs_backend.write_stdout(output)
        self._vs_backend.send_command_executed()
```

After doing that, reset the interactive window in VS and it should work fine.

I don't think we'll be able to get this fix into 2.2 at this stage. It's a regression due to an external product, but as we are so close to releasing we are being **extremely** careful about making any changes at all. There'll be some discussions within the team, no doubt, but I think this is unlikely.
